src/thing/build.test.ts

```diff
@@ -0,0 +1,217 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import { describe, it, expect } from "@jest/globals";
+import { buildThing } from "./build";
+import * as adders from "./add";
+import * as setters from "./set";
+import * as removers from "./remove";
+import { asIri, createThing, isThing } from "./thing";
+import { IriString, ThingLocal, ThingPersisted } from "../interfaces";
+
+describe("Thing Builder API", () => {
+  it("creates the same Thing as the non-fluent functions", () => {
+    const startingThing = createThing();
+
+    const builtThing = buildThing(startingThing)
+      .addInteger("https://some.vocab/predicate", 42)
+      .addInteger("https://some.vocab/predicate", 1337)
+      .addStringWithLocale(
+        "https://some.vocab/predicate",
+        "Some string",
+        "nl-nl"
+      )
+      .setStringWithLocale(
+        "https://some.vocab/predicate",
+        "Some other string",
+        "nl-nl"
+      )
+      .addStringWithLocale(
+        "https://some.vocab/predicate",
+        "Yet another string",
+        "en-gb"
+      )
+      .removeStringWithLocale(
+        "https://some.vocab/predicate",
+        "Yet another string",
+        "en-gb"
+      )
+      .removeInteger("https://some.vocab/predicate", 1337)
+      .addBoolean("https://some-other.vocab/predicate", true)
+      .setBoolean("https://some-other.vocab/predicate", false)
+      .removeAll("https://some-other.vocab/predicate")
+      .build();
+
+    let nonBuilderThing = adders.addInteger(
+      startingThing,
+      "https://some.vocab/predicate",
+      42
+    );
+    nonBuilderThing = adders.addInteger(
+      nonBuilderThing,
+      "https://some.vocab/predicate",
+      1337
+    );
+    nonBuilderThing = adders.addStringWithLocale(
+      nonBuilderThing,
+      "https://some.vocab/predicate",
+      "Some string",
+      "nl-nl"
+    );
+    nonBuilderThing = setters.setStringWithLocale(
+      nonBuilderThing,
+      "https://some.vocab/predicate",
+      "Some other string",
+      "nl-nl"
+    );
+    nonBuilderThing = removers.removeInteger(
+      nonBuilderThing,
+      "https://some.vocab/predicate",
+      1337
+    );
+    nonBuilderThing = adders.addStringWithLocale(
+      nonBuilderThing,
+      "https://some.vocab/predicate",
+      "Yet another string",
+      "en-gb"
+    );
+    nonBuilderThing = removers.removeStringWithLocale(
```

Comment:
I'm not sure why all these calls to `remove` and `set` - isn't this test just asserting that the builder builds a dataset as expected?

Reply:
Agreed - this test looks like it's trying to cover a _lot_ of functionality. Each function should have its own test, or failures will be difficult to diagnose - or potentially hidden by all of the functions being called.

Reply:
Yeah, I initially just wanted a relatively small test to verify that it did the same as one of them, since it's practically trivial that if one works, the rest will too. But then of course it was complaining about incomplete coverage, so I just added a couple of other calls.

Split up in 3f0f83c2.